### PR TITLE
Gh-341: Adds dedicated cardinality docs

### LINKED ISSUE #341: Clearly document Gaffer Cardinality and HLL
The current documentation on HLL is very developer oriented, there should be some user docs added that clearly explain how to:
- Add a HLL Cardinality property to a graph
- Add Entities with cardinality property set

### PR DESCRIPTION
**Pull request by t92549**

### CODE REVIEW

File: docs/getting-started/guide/cardinality.md [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+There are two main methods of adding cardinality elements in Gaffer. The first is to do it manually when you add your edges, the second is to use a generator that can do it for you.
+
+### Manually adding cardinality entities
+
+Adding a cardinality entity between vertex 'A' and 'B' using `AddElements`. For more examples of different property types, see the skectehs [dev docs](../../dev/rest-api-sketches.md#primitive-data-types-over-json) page.
+
+=== "Java"
+
+    ``` java
+    final HllSketch hll = new HllSketch(10); // (1)
```
Comment (t92549) [inline docs/getting-started/guide/cardinality.md]: Unfortunately, this comment annotation doesn't seem to work when inside the `=== "Java"`, do you know anything about this? @GCHQDeveloper314

Comment (t92549) [reply docs/getting-started/guide/cardinality.md]: Fixed it

File: docs/reference/properties-guide/map-set.md [hunk 2]
```diff
@@ -78,7 +78,7 @@ Property | Serialiser
...
+- BoundedTimestampSet: This is similar to the above set, except that when it is created a maximum size is specified. If the set grows beyond the maximum size, then a random sample of the timestamps of that size is created. This is useful to avoid the size of the set for entities or edges that are very busy growing too large. By retaining a sample, we maintain an indication of when the entity or edge was active, without retaining all the information. The sample is implemented using a [ReservoirLongsUnion](https://github.com/apache/datasketches-java/blob/4.0.0/src/main/java/org/apache/datasketches/sampling/ReservoirLongsSketch.java) from the Datasketches library.
```
Comment (GCHQDeveloper314) [inline docs/reference/properties-guide/map-set.md]: You've changed the file linked to be `ReservoirLongsSketch`, but the link still says `ReservoirLongsUnion`.

File: docs/reference/properties-guide/advanced.md [hunk 3]
```diff
@@ -4,26 +4,26 @@ These properties use advanced sketch structures from the Clearspring and Dataske
...
 
-- Using a HyperLogLogPlusPlus sketch to provide a very quick estimate of the degree of a node.
+- Using a Hll sketch to provide a very quick estimate of the degree of a node.
 - Using a quantiles sketch to estimate the median score associated to an edge, or the 99th percentile of the scores seen on an edge.
 - Using a reservoir items sketch to store a sample of all the distinct labels associated to an edge.
 - Using theta sketches to estimate the number of distinct edges seen on a particular day, the number seen on the previous day and the overlap between the two days.
 
-Gaffer provides serialisers and aggregators for sketches from two different libraries: the [Clearspring](https://github.com/addthis/stream-lib) library and the pre-Apache version of the [Datasketches](https://datasketches.github.io/) library.
+Gaffer provides serialisers and aggregators for sketches from two different libraries: the Apache version of the [Datasketches](https://datasketches.apache.org/) library and the [Clearspring](https://github.com/addthis/stream-lib) library which has been **deprecated in Gaffer**.
```
Comment (GCHQDeveloper314) [inline docs/reference/properties-guide/advanced.md]: I think the info on deprecation could go into a separate section which go on to could explain that Clearspring is no longer maintained.

File: mkdocs.yml [hunk 4]
```diff
@@ -59,12 +59,13 @@ nav:
       - 'Gaffer Guide':
         - 'Gaffer Guide': 'getting-started/guide/guide.md'
         - 'CSV Import and Export': 'getting-started/guide/csv.md'
-        - 'Sketches with the REST API': 'getting-started/guide/rest-api-sketches.md'
+        - 'Cardinality': 'getting-started/guide/cardinality.md'
```
Comment (GCHQDeveloper314) [inline mkdocs.yml:62]: Cardinality is ok for the file name but page name could be more informative.

File: docs/getting-started/guide/cardinality.md [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+  1 --> 5
+  3(3, cardinality=2) --> 5
+  3 --> 5
+  3 --> 5
+  3 --> 5
+```
+
+For very large graphs, updating this number accurately would be very costly in compute. This is because for each new Edge that is added, we would have to check both connected Entities to see if they are already connected to the other Entity, and this could be costly for Entities with a high cardinality. Instead, Gaffer uses approximate cardinality making use of a [HyperLogLog Sketch](https://datasketches.apache.org/docs/HLL/HLL.html), which estimates the cardinality with relatively low error.
+
+## How to add cardinality to your graph
```
Comment (GCHQDeveloper314) [inline docs/getting-started/guide/cardinality.md:23]: This page refers to cardinality estimates as just cardinality. I think it would be good for it to mention every now and then that this is an estimate. It might also be good to very briefly explain how to add the real (not estimate) cardinality as well or why that can't be done.

File: docs/dev/managing-dependencies.md [hunk 6]
```diff
@@ -38,13 +38,13 @@ Here's a simplified tree output with the verbose output explained (expand the :m
 uk.gov.gchq.gaffer:core:pom:2.0.0
 +- org.slf4j:slf4j-api:jar:1.7.36:compile
 +- org.slf4j:slf4j-reload4j:jar:1.7.36:compile
-|  +- (org.slf4j:slf4j-api:jar:1.7.36:compile - omitted for duplicate) // (1)
-|  \- ch.qos.reload4j:reload4j:jar:1.2.18.3:compile (version managed from 1.2.19) // (2)
+|  +- (org.slf4j:slf4j-api:jar:1.7.36:compile - omitted for duplicate) // (1)!
```
Comment (GCHQDeveloper314) [inline docs/dev/managing-dependencies.md:41]: Why this change?

Comment (t92549) [reply docs/dev/managing-dependencies.md:41]: Adding the `!` removes the comment from the code sample and looks cleaner

File: mkdocs.yml (hunk 4 above)
Comment (t92549) [reply mkdocs.yml:62]: Any suggestion? I like the simplicity of just "Cardinality"

File: docs/reference/properties-guide/advanced.md (hunk 3 above)
Comment (t92549) [reply docs/reference/properties-guide/advanced.md]: The reasoning is in the [dev docs ](https://github.com/gchq/gaffer-doc/pull/351/files/eec28dfa46b7c41b22e72360cfbc8c14b5d09c98#diff-6a02a5a66873f2f772a9328dd0a089dd63bd83dd8fcf917bfcf86f9830af0860R20)

File: docs/getting-started/guide/cardinality.md (hunk 5 above)
Comment (t92549) [reply docs/getting-started/guide/cardinality.md:23]: It is mentioned why it shouldn't be done: "For very large graphs, updating this number accurately would be very costly in compute. This is because for each new Edge that is added, we would have to check both connected Entities to see if they are already connected to the other Entity, and this could be costly for Entities with a high cardinality. Instead, Gaffer uses approximate cardinality..."

I am not even really sure it is possible right now in Gaffer to store the "real" cardinality in the same way due to lack of aggregators/serialisers for a Set of objects which only returns its length. Instead you would just store a set of vertices and then get the length. For this reason, approximate cardinality is always used in Gaffer and the Entity group is often just called "cardinality".

File: docs/reference/properties-guide/advanced.md (hunk 3 above)
Comment (GCHQDeveloper314) [reply docs/reference/properties-guide/advanced.md]: That reasoning in the dev docs doesn't explain that Clearspring is no longer maintained. We don't need repeat the reasoning though, as we can link to the devs docs from here.

File: docs/dev/managing-dependencies.md (hunk 6 above)
Comment (GCHQDeveloper314) [reply docs/dev/managing-dependencies.md:41]: Ok, [this is the doc link for reference](https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#stripping-comments).

File: docs/getting-started/guide/cardinality.md [hunk 7]
```diff
@@ -0,0 +1,274 @@
+# Cardinality
+
+This page will describe what cardinality is, how to add it to your Gaffer graph and how to use it in Gaffer queries.
+
+## What is Cardinality?
+
+Cardinality represents the number of unique Edges that are connected to a given Entity, also known as the degree of an Entity.
```
Comment (GCHQDeveloper314) [inline docs/getting-started/guide/cardinality.md]: Maybe this is what Gaffer means by Cardinality but this not compatible with graph theory definitions. Both the explanation for Cardinality and degree are incorrect in that sense. This needs to be expanded to clarify how Gaffer's use of these terms differ from the norm.

This is already a major improvement on [the old docs](https://gchq.github.io/gaffer-doc/v1docs/getting-started/user-guide/cardinalities.html) which don't mention unique connections at all!

File: docs/getting-started/guide/cardinality.md (hunk 5 above)
Comment (GCHQDeveloper314) [reply docs/getting-started/guide/cardinality.md:23]: > We would have to check both connected Entities to see if they are already connected to the other Entity.

Only if we wanted the number of unique connections. If we didn't then a simple count property could be used which was just iterated whenever an edge was added. This is why we want to use HLLP, but I missed this because cardinality is not intrinsically a count of unique connections, but in Gaffer it is.

**Review by GCHQDeveloper314 — CHANGES_REQUESTED** [review]
Having taken a closer look at the definitions for some of the language which is used on the 'Cardinality' page. (e.g. [Cardinality](https://en.wikipedia.org/wiki/Cardinality), [Degree](https://en.wikipedia.org/wiki/Degree_(graph_theory)) & [HLL](https://en.wikipedia.org/wiki/HyperLogLog)). I think we need to be very clear about what Gaffer means with these words.

If I wrote this myself I wouldn't use 'Cardinality' at all; 'Distinct Connections', 'Unique Connections' or 'count-distinct' would be more precise. I don't propose this is changed, but we could mention that the property can be called whatever the user wants if they'd prefer not to use 'Cardinality'.

HLL and other sketches help with the [count-distinct problem](https://en.wikipedia.org/wiki/Count-distinct_problem), also known as the cardinality estimation problem. This is probably why this was always called Cardinality in Gaffer. The mathematical cardinality of a set with repeated items doesn't count the duplicates. However in Gaffer items (edges) may have different properties (distinct) but the same connected entities. So the edges in the set are not repeated as with a mathematical set. This is why the `HllSketch` takes in the destination entity only.

File: docs/getting-started/guide/cardinality.md (hunk 7 above)
Comment (t92549) [reply docs/getting-started/guide/cardinality.md]: Have made some changes, any better?

File: docs/reference/properties-guide/advanced.md (hunk 3 above)
Comment (t92549) [reply docs/reference/properties-guide/advanced.md]: Is this okay now?

**Review by GCHQDeveloper314 — APPROVED** [review]
